File: 2020-04/4-20/98验证二叉搜索树/98.py
#  -*- coding: utf-8 -*-

#  Editor      : Pycharm
#  File        : 98.py
#  Author      : ZhenyuMa
#  Created     : 2020/4/21 12:15 上午
#  Description : 小递归


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def isValidBST(self, root: TreeNode) -> bool:
        return self.helper(root, None, None)

    def helper(self, root, lower, upper):
        if root is None:
            return True
        if lower is not None and root.val <= lower:
            return False
        if upper is not None and root.val >= upper:
            return False
        return self.helper(root.left, lower, root.val) and \
               self.helper(root.right, root.val, upper)

File: 2020-05/5-05/98验证二叉搜索树/98.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool isValidBST(TreeNode* root) {
        if(!root)
            return true;
        
        return isValidBSTRoot(root, nullptr, nullptr);
    }

    bool isValidBSTRoot(TreeNode* root, TreeNode* lower, TreeNode* upper){
        if(!root)
            return true;
        
        if(lower && root->val <= lower->val)
            return false;
        if(upper && root->val >= upper->val)
            return false;

        return isValidBSTRoot(root->left, lower, root) && isValidBSTRoot(root->right, root, upper);
    }
};